src/sentry/sentry_metrics/querying/visitors/query_expression.py

```diff
@@ -290,78 +290,124 @@ class UnitsNormalizationVisitor(QueryExpressionVisitor[QueryExpression]):
 
     def __init__(self):
         self._unit_family = None
-        self._reference_unit = None
-        self._scaling_factor = None
-
-        self._is_formula = False
 
-    def _visit_formula(self, formula: Formula) -> QueryExpression:
-        self._is_formula = True
+    def _visit_formula(self, formula: Formula) -> tuple[UnitMetadata, QueryExpression]:
+        last_metadata: WithUnit | None = None
+        future_units = []
 
         has_all_timeseries_params = True
+        has_all_futures = True
+
         parameters = []
-        for parameter in formula.parameters:
+        for index, parameter in enumerate(formula.parameters):
             if not isinstance(parameter, Timeseries):
                 has_all_timeseries_params = False
 
-            parameters.append(self.visit(parameter))
-
-        # If we have all timeseries as parameters of a formula and the function is belonging to `*` or `/` we will
-        # not perform any units normalization.
-        # TODO: we might want to implement units normalization following a more mathematical approach like `ms^2` or
-        #  `byte/s` but this is going to come at a later point.
+            unit_metadata, query_expression = self.visit(parameter)
+            if isinstance(unit_metadata, WithNoUnit):
+                return unit_metadata, formula
+            elif isinstance(unit_metadata, WithFutureUnit):
```

Comment:
Future units are those units that have not been normalized when `self.visit` was called and they must be normalized only when we know the unit family and reference unit of the formula itself. Numeric scalars are one example, since they need to be normalized in the context in which they are used.